### PR TITLE
Feature/api

### PR DESCRIPTION
**Pull request by art-vasilyev**
## Description

* low-level API implementation
* use low-level API for API requests from client API
* use pytest and pytest-vcr for tests
* drop python 2.7 support

## Checklist

- [X] Tests (if applicable)
- [x] Documentation (if applicable)
- [x] Changelog stub (or use [conventional commit messages](https://www.conventionalcommits.org/))

### CODE REVIEW

File: pyuploadcare/resources/file_group.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        if isinstance(key, slice):
+            raise TypeError("slicing is not supported")
+        else:
+            file_info = self.info()["files"][key]
+            if file_info is not None:
+                return File.construct_from(file_info)
+
+    @property
+    def _api_uri(self):
+        return "groups/{0}/".format(self.id)
```
Comment (dmitry-mukhin) [inline pyuploadcare/resources/file_group.py:86]: here and below switch to f string for consistency

File: pyuploadcare/resources/file_group.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        file_cdn_urls = []
+        for file_index in range(len(self)):
+            file_cdn_url = "{group_cdn_url}nth/{file_index}/".format(
+                group_cdn_url=self.cdn_url, file_index=file_index
+            )
+            file_cdn_urls.append(file_cdn_url)
+        return file_cdn_urls
+
+    def info(self):
```
Comment (dmitry-mukhin) [inline pyuploadcare/resources/file_group.py:126]: also, switch to properties?

File: tests/integration/test_api_resources.py [hunk 3]
```diff
@@ -1,379 +1,320 @@
...
 
-        while time.time() - time_started < timeout:
-            status = file_from_url.update_info()['status']
-            if status in ('success', 'failed', 'error'):
-                break
-            time.sleep(1)
+    assert isinstance(file_from_url.get_file(), File)
+
+
+@skip_on_travis
```
Comment (dmitry-mukhin) [inline tests/integration/test_api_resources.py:65]: rename to `skip_on_ci`

File: tests/integration/utils.py [hunk 4]
```diff
@@ -29,5 +26,6 @@ def create_file_group(files_qty=1):
 
 # travis-ci.org runs our tests in an overworked virtual machine, which makes
 # timing-related tests unreliable.
-skip_on_travis = unittest.skipIf('TRAVIS' in os.environ,
-                                 'timing tests unreliable on travis')
+skip_on_travis = unittest.skipIf(
+    "TRAVIS" in os.environ, "timing tests unreliable on travis"
+)
```
Comment (dmitry-mukhin) [inline tests/integration/utils.py:31]: we're not using travis anymore.
is this still a problem in GHA?
if so, rename to `skip_on_ci`

**Review by dmitry-mukhin — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -85,42 +80,39 @@ jobs:
...
-        pip install -e .
-    - name: Install mock on py2.7
-      if: ${{ matrix.python-version == '2.7' }}
-      run: |
-        pip install mock
+        poetry install
     - name: Test with pytest
       run: |
-        pytest tests/dj --cov=pyuploadcare
+        poetry run pytest tests/dj --cov=pyuploadcare
```
[the remark this answers is not in the document]
Comment (art-vasilyev) [reply .github/workflows/test.yml]: done

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -85,42 +80,39 @@ jobs:
...
-        pip install -e .
-    - name: Install mock on py2.7
-      if: ${{ matrix.python-version == '2.7' }}
-      run: |
-        pip install mock
+        poetry install
     - name: Test with pytest
       run: |
-        pytest tests/integration --cov=pyuploadcare
+        poetry run pytest tests/integration --cov=pyuploadcare
```
[the remark this answers is not in the document]
Comment (art-vasilyev) [reply .github/workflows/test.yml]: done

File: poetry.lock [hunk 7]
```diff
@@ -0,0 +1,1698 @@
+[[package]]
```
[the remark this answers is not in the document]
Comment (art-vasilyev) [reply poetry.lock]: libraries do not actually need to lock dependencies, removed

File: pyuploadcare/resources/file.py [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+            >>> file_.default_effects = 'effect/flip/-/effect/mirror/'
+            >>> file_.cdn_url
+            https://ucarecdn.com/a771f854-c2cb-408a-8c36-71af77811f3b/-/effect/flip/-/effect/mirror/
+
+        """
+        return "{cdn_base}{path}".format(
+            cdn_base=conf.cdn_base, path=self.cdn_path(self.default_effects)
+        )
+
+    def info(self):
```
[the remark this answers is not in the document]
Comment (art-vasilyev) [reply pyuploadcare/resources/file.py:128]: done

File: pyuploadcare/resources/file_group.py (hunk 1 above)
Comment (art-vasilyev) [reply pyuploadcare/resources/file_group.py]: done

File: pyuploadcare/resources/file_group.py (hunk 2 above)
Comment (art-vasilyev) [reply pyuploadcare/resources/file_group.py:115]: done

File: tests/integration/utils.py (hunk 4 above)
Comment (art-vasilyev) [reply tests/integration/utils.py]: didn't notice any problems with tests on CI, removed

File: tests/integration/test_api_resources.py (hunk 3 above)
Comment (art-vasilyev) [reply tests/integration/test_api_resources.py]: didn't notice any problems with tests on CI, removed

**Comment by dmitry-mukhin** [comment]
We have to add a note in readme or somewhere else, that migrating from v2 to v3 involves switching from functions to properties in some cases.

**Comment by dmitry-mukhin** [comment]
otherwise, shipit

**Review by dmitry-mukhin — APPROVED** [review]

**Pull request merged**